### PR TITLE
Defines new API of Model

### PR DESCRIPTION
**Pull request by euph00**
I have no doubt in my heart that this is buggy as hell as it stands. Will be adding tests and doing debugging along the way tomorrow. 

This does not compile as it stands (rightfully so, it wouldn't make sense if it could compile now)

However, as far as the other components are concerned, the API being there is good enough to start preliminary work in refactoring other components to fit the new API, even though actual testing will require a debugged implementation of Model. Please refer to Model.java interface to see the important changes. The other changes are all implementation changes within the model package and are not meant to be accessed from outside model package.

Will update again when it is debugged. For now, please review this API for sanity check. Refer to diff to see changes.

### CODE REVIEW

**Review by pzhengze — COMMENTED** [review]
Looks alright but what is your plan for saving to storage? Because I don't see anything about that so far. Right now the storage configured for saving persons so are you going to change it to wards? Or are you going to create a JsonAdaptedWard class? Then how are you going to link the wards and the persons?

**Comment by pzhengze** [comment]
Could possibly have a ward json that contains the patients details then while we read the ward json we call the SsonAdaptedPerson make the persons. Something like how the JsonAdaptedPerson calls JsonAdaptedTag

File: src/main/java/seedu/patientist/model/person/patient/PatientIdNumber.java [hunk 1]
```diff
@@ -1,16 +1,32 @@
...
 
+import static java.util.Objects.requireNonNull;
+import static seedu.patientist.commons.util.AppUtil.checkArgument;
+
 /**
  * Object representing the ID number of a patient. ID numbers are automatically capitalised when they are created.
  */
 public class PatientIdNumber {
+
+    public static final String VALIDATION_REGEX = "\\p{Alnum}+";
```
[the remark this answers is not in the document]
Comment (euph00) [reply src/main/java/seedu/patientist/model/person/patient/PatientIdNumber.java:11]: I came up with those examples arbitrarily, and in fact this is just an arbitrary check as well. PID isnt based off a real thing anyway, so I just left it as this for now.

[the remark this answers is not in the document]
Comment (pzhengze) [reply src/main/java/seedu/patientist/model/person/patient/PatientIdNumber.java:11]: Alright. We'll leave it like this for now.

**Review by lukkesreysandeur — COMMENTED** [review]
I'm okay with the changed being made. We could just add on using this base API once merged, but I expect that we may need to make a bunch more changes to get the app working again.

**Comment by euph00** [comment]
> Could possibly have a ward json and JsonAdaptedWard that contains the patients details then while we read the ward json we call the SsonAdaptedPerson make the persons. Something like how the JsonAdaptedPerson calls JsonAdaptedTag currently

The implementation of saving to storage (JsonAdaptedWard) etc is up to whoever is doing the storage component. Model just defines the structure of data and an API to interact with the data, how it is parsed to be written to storage is abstracted away from Model.

**Pull request merged**